src/main/java/com/atlassian/db/replica/api/ThrottledCache.java

```diff
@@ -57,36 +62,106 @@ private ThrottledCacheWithTimeout(Clock clock, Duration timeout) {
         @Override
         public Optional<T> get(Supplier<T> supplier) {
             maybeRefresh(supplier);
-            return Optional.ofNullable(value);
+            return Optional.ofNullable(lockValuePair.get().getValue());
         }
 
         @Override
         public Optional<T> get() {
-            return Optional.ofNullable(value);
+            return Optional.ofNullable(lockValuePair.get().getValue());
         }
 
         private void maybeRefresh(Supplier<T> supplier) {
             final Instant deadline = clock.instant().plus(this.timeout);
-            final boolean hasLock = lock.compareAndSet(null, deadline);
+            final T currentValue = this.lockValuePair.get().getValue();
+            final LockValuePair<T> valueWithLock = new LockValuePair<>(
+                deadline,
+                currentValue
+            );
+            final LockValuePair<T> currentLockValuePair = this.lockValuePair.get();
+            final boolean hasLock = !currentLockValuePair.isLocked() && this.lockValuePair.compareAndSet(
+                currentLockValuePair,
+                valueWithLock
+            );
             if (hasLock) {
-                loadValue(supplier, deadline);
+                loadValue(supplier, valueWithLock);
             } else {
-                final Instant instant = lock.get();
-                final boolean didLockExpire = instant != null && instant.isBefore(clock.instant());
-                if (didLockExpire) {
-                    final boolean didLockTakeOver = lock.compareAndSet(instant, deadline);
+                final LockValuePair<T> oldLock = this.lockValuePair.get();
+                if (oldLock.didLockExpire(clock)) {
+                    final LockValuePair<T> newLock = new LockValuePair<>(
+                        deadline,
+                        oldLock.getValue()
+                    );
+                    final boolean didLockTakeOver = lockValuePair
+                        .compareAndSet(oldLock, newLock);
                     if (didLockTakeOver) {
-                        loadValue(supplier, deadline);
+                        loadValue(supplier, newLock);
                     }
                 }
             }
         }
 
-        private void loadValue(Supplier<T> supplier, Instant timeout) {
+        private void loadValue(Supplier<T> supplier, LockValuePair<T> currentLockValue) {
             try {
-                value = supplier.get();
-            } finally {
-                lock.compareAndSet(timeout, null);
+                final T newValue = supplier.get();
+                this.lockValuePair.compareAndSet(
+                    currentLockValue,
+                    new LockValuePair<>(
+                        null,
```

Comment:
why null, not `now` ?

Reply:
null -> lock released
now() -> lock expired